py/kubeflow/kubeflow/ci/update_jupyter_web_app.py

```diff
@@ -0,0 +1,269 @@
+      with open(os.path.join(os.getenv("HOME"), ".ssh", "known_hosts"),
+                mode='a') as hf:
+        hf.write(output)
+
+    if not remote_fork.startswith("git@github.com"):
+      raise ValueError("Remote fork currently only supports ssh")
+
+    remote_repo = self._find_remote_repo(repo, remote_fork)
+
+    if not remote_repo:
+      fork_name = remote_fork.split(":", 1)[-1].split("/", 1)[0]
+      logging.info("Adding remote %s=%s", fork_name, remote_fork)
+      remote_repo = repo.create_remote(fork_name, remote_fork)
+
+    logging.info("Last change to components-jupyter-web-app was %s", last_commit)
+
+    base = "gcr.io/{0}/jupyter-web-app".format(registry_project)
+
+    # Check if there is already an image tagged with this commit.
+    image = base + ":" + self.last_commit
+    transport = transport_pool.Http(httplib2.Http)
+    src = docker_name.from_string(image)
+    creds = docker_creds.DefaultKeychain.Resolve(src)
+
+    image_exists = False
+    try:
+      with v2_2_image.FromRegistry(src, creds, transport) as src_image:
+        logging.info("Image %s exists; digest: %s", image,
+                     src_image.digest())
+        image_exists = True
+    except docker_http.V2DiagnosticException as e:
+      if e.status == 404:
+        logging.info("%s doesn't exist", image)
+      else:
+        raise
+
+    if not image_exists:
+      logging.info("Building the image")
+      image = self.build_image(build_project, registry_project)
+    else:
+      logging.info("Image %s already exists", image)
+
+    # TODO(jlewi):We should check what the current image and not update it
+    # if its the existing image
+    prototype_file = self.update_prototype(image)
+
+    branch_name = "update_jupyter_{0}".format(last_commit)
+
+    if repo.active_branch.name != branch_name:
+      logging.info("Creating branch %s", branch_name)
+      util.run(["git", "checkout", "-b", branch_name], cwd=self._root_dir())
+
+    logging.info("Add file %s to repo", prototype_file)
+    repo.index.add([prototype_file])
+    repo.index.commit("Update the jupyter web app image to {0}".format(image))
+
+    util.run(["git", "push", "-f", remote_repo.name], cwd=self._root_dir())
+
+    self.create_pull_request(commit=last_commit)
+
+  def create_pull_request(self, base="kubeflow:master", commit=None):
+    """Create a pull request.
+
+    Args:
+      base: The base to use. Defaults to "kubeflow:master". This should be
+        in the form <GitHub OWNER>:<branch>
+    """
+    # TODO(jlewi): Modeled on
+    # https://github.com/kubeflow/examples/blob/master/code_search/docker/ks/update_index.sh
+    # TODO(jlewi): We should use the GitHub API and check if there is an
+    # existing open pull request. Or potentially just use the hub CLI.
+
+    if not commit:
+      commit = self.last_commit
+      logging.info("No commit specified defaulting to %s", commit)
+
+    pr_title = "Update the jupyter-web-app image to {0}".format(commit)
```

Comment:
Add something like "Automated PR to update..." in the title.

Reply:
Done